Adding closable keyword option to display_flash_messages

jeffse opened the pull request:
Adding the optional keyword option 'closable' to the
display_flash_messages method that accepts a boolean true/false,
defaulting to true.
If false, suppresses the close buttons on the callouts.

jeffse commented:
I wanted to be able to use the display_flash_messages but found the close buttons to be not quite right for my project. This adds the 'closable' option to the method.

dsandstrom commented:
Thanks, looks good.

dgmstuart commented:
Hi @jeffse - the foundation-6 branch has been cleaned up a bit - could you rebase on top of it?

dgmstuart commented:
Context: here are the docs: http://foundation.zurb.com/sites/docs/close-button.html

dgmstuart commented on lib/foundation_rails_helper/flash_helper.rb:
```diff
@@ -18,29 +18,29 @@ module FlashHelper
-    def display_flash_messages(key_matching = {})
-      key_matching = DEFAULT_KEY_MATCHING.merge(key_matching)
+    def display_flash_messages(**options)
+      closable = true
```
Do we definitely want to default this to true?

dgmstuart commented on lib/foundation_rails_helper/flash_helper.rb:
```diff
@@ -18,29 +18,29 @@ module FlashHelper
       :warning   => :warning,
       :primary   => :primary
     }
-    def display_flash_messages(key_matching = {})
```
Note that the example in the comment above already includes the closable attributes, but doesn't quite match what is in the foundation 6 docs: http://foundation.zurb.com/sites/docs/close-button.html

dgmstuart commented on lib/foundation_rails_helper/flash_helper.rb:
```diff
@@ -18,29 +18,29 @@ module FlashHelper
     }
-    def display_flash_messages(key_matching = {})
-      key_matching = DEFAULT_KEY_MATCHING.merge(key_matching)
+    def display_flash_messages(**options)
```
Could we have closable as a second argument instead? Maybe a keyword argument?

`key_matching` is a lookup, not an options hash, so it doesn't make sense to me to pass `:closable` as part of the existing hash.

I think the best implementation would be to use keyword arguments for both - something like:

```ruby
def display_flash_messages(closable: true, key_matching: {})
```

That would be a breaking change, but then again moving to foundation 6 is a breaking change, so I don't see that as a reason not to do it. Maybe we'd need to add some upgrade notes.

An alternative would be an options hash, which is essentially the same, but a little more verbose - something like:

```ruby
def display_flash_messages(options: {})
  closable = options.fetch(:closable, true)
  new_key_matching = options[:key_matching] || {}
  key_matching = DEFAULT_KEY_MATCHING.merge(new_key_matching)
```

dsandstrom replied:
I like the second option using one options hash.  That way, we can easily add more options, if need be, besides `closable`.

dgmstuart replied:
@dsandstrom well we can always add more keyword arguments as well ;) 
Either way the body of the method needs to know about them, so using an options hash wouldn't decouple anything.

jeffse replied:
I originally wanted to set `closable` as a keyword argument just as @dgmstuart suggests above, but as you mention it would be a breaking change. How many other breaking changes are there in the Foundation 6 branch?

dgmstuart replied:
@jeffse forget I said anything: adding foundation 6 is a fundamentally breaking change so we should feel free to change other parts of the api where it makes sense.

The only thing I'd say is, since this is a potentially subtle breaking change (potentially a button might end up not being the expected colour), we should mention this somewhere - at least in the changelog. 

Or maybe we just need to flag it up to @potomak, who added the feature of having custom hashes in the first place and may be the only person using this feature 😜

potomak replied:
@dgmstuart breaking changes are fine as long as the major version number is updated. Note: I don't actively use this gem anymore.

dgmstuart commented on lib/foundation_rails_helper/flash_helper.rb:
```diff
@@ -18,29 +18,29 @@ module FlashHelper
+    def alert_box(value, alert_class, closable)
+      options = { class: "flash callout #{alert_class}" }
+      options[:data] = { closable: '' } if closable
+      content_tag(:div, options) do
```
Putting these in an `options` hash is a nice touch 😊 👍

dgmstuart commented on spec/foundation_rails_helper/flash_helper_spec.rb:
```diff
@@ -10,6 +10,8 @@
   include ActionView::Helpers::FormTagHelper
   include FoundationRailsHelper::FlashHelper
 
+  RSpec::Matchers.define_negated_matcher :not_have_css, :have_css
```
Is this necessary? Could we use the existing negated matcher `have_no_css` or even `to_not have_css`?

jeffse replied:
Aha, I was not aware of the `have_no_css` matcher. Good to know!

dsandstrom commented on spec/foundation_rails_helper/flash_helper_spec.rb:
```diff
@@ -83,4 +85,15 @@
+      node = Capybara.string display_flash_messages(closable: false)
+      expect(node).
+        to  have_css("div.flash.callout.success", :text => "Flash message").
+        and not_have_css("[data-close]", :text => "×")
```
TIL about [compound expectations](https://www.relishapp.com/rspec/rspec-expectations/v/3-5/docs/compound-expectations)

dgmstuart replied:
Ooh, now that I see this, the normal formatting of this is with the dots at the start:

```ruby
expect(node)
  .to have_css("div.flash.callout.success", :text => "Flash message")
  .and not_have_css("[data-close]", :text => "×")
```

dgmstuart commented on .rubocop.yml:
```diff
@@ -0,0 +1,17 @@
+inherit_from: .rubocop_todo.yml
```
@jeffse did you mean to add these onto your local master instead of the branch?

We need to get this PR into a state where it only has your commits on it - let me know if I can help with that at all.

jeffse replied:
Yeah, I'm having some trouble getting this into a usable state, integrating with the latest commits from the foundation-6 branch. I may just scrap this fork and start over, and given @potomak's comments from today, likely refactor this to be a breaking change to display_flash_messages.

Probably have to wait until tomorrow though. My kids are home from school early today due to a broken water main at the school. No toilets, no school!

dgmstuart replied:
@jeffse no need to nuke and start again - this is all fixable. 
The steps are: 

1. Update your local master branch to be in line with the master branch on this repo 
2. Update your local copy of the foundation-6 branch to be up to date with this repo
3. Rebase your pull request branch on top of the foundation-6 branch and push it back to this PR

Here is the help page on syncing your fork: https://help.github.com/articles/syncing-a-fork/

... but I'll break it down for you, assuming only basic git experience (apologies if that's too low-level):

1. First, make sure you have a branch pointing to this repo: 

  ```
  git remote add upstream https://github.com/sgruhier/foundation_rails_helper
  ```
2. Now update your version of the master branch:
  
  ```
  git checkout master
  git merge upstream/master
  git push
  ```

3. Now update your version of the foundation-6 branch (delete it and fetch it again). You'll probably have to force push when you push:

  ```
  git branch -D foundation-6
  git fetch upstream
  git checkout -b foundation-6 upstream/foundation-6
  git push --force-with-lease
  ```

4. Now rebase your feature on top of the foundation-6 branch. This might seem a bit hairy if you haven't done an interactive rebase before, but there's nothing to be concerned about:

  ```
  git checkout foundation-6-closable-option
  git rebase --interactive foundation-6
  ```

  This should open your editor. You'll want to delete all of the lines except the ones relating to commits which you've created (I think there's only one?). When you save and close that file, the rebase should happen. You might get some conflicts relating to formatting (using the old hash style, using double quotes etc.) but hopefully nothing else. 

5. Once you're happy with your local branch, you can force-push it back to your fork

  ```
  git push origin foundation-6-closable-option --force-with-lease
  ```

### Rescuing from screwups

* If anything goes wrong during the rebase, you can quit it with `git rebase abort`
* If your local state gets messed up, but you haven't pushed anything you can reset it back to the state on your github with e.g. `git reset origin/foundation-6-closable-option`

Let me know how you get on, and I'm very happy to help if you have any problems or questions.

jeffse commented:
Thanks for your git help, @dgmstuart! I'm still a relative newbie to git, but I think I was able to finally resolve everything here.

I changed display_flash_messages to use keyword arguments, and added some documentation in the comments of the method at least.

dgmstuart commented:
@jeffse looks like we're getting closer, but there are now 26 commits on this branch - most of which are previous commits by other people. We need this branch to ONLY contain your commits which are for this feature - I think it's only these two: 

  * Adding closable keyword option to display_flash_messages
  * Changing display_flash_messages to use keyword arguments

...and actually you're probably better off squashing those two into one. 

The step in my instructions which is supposed to get rid of all the other commits is this one, during the rebase:

>You'll want to delete all of the lines except the ones relating to commits which you've created 

did you try that?

dgmstuart commented:
@jeffse actually it looks like your `master` and `foundation-6` branches are way behind the ones on this repo - you'll need to follow my instructions from the beginning to get them up-to-date.

jeffse commented:
I already spent two hours doing a bunch of manual merges from what I thought were correct. I really think it's better for me to trash this branch, fork off a brand-new copy, and make my changes there.

dgmstuart commented:
@jeffse sorry to hear that. Best of luck with your new branch.
